### PR TITLE
[JSONRPC] Expose all players to Player.GetPlayers and add game media type

### PR DESCRIPTION
**Pull request by enen92**
## Description
This PR allows to get all configured players using `Player.GetPlayers` instead of only players that supports audio or video as before. `CPlayerCoreFactory::GetPlayers()` was filtering internally the list of configured players by audio or video - and as a result Retroplayer was missing. Took the opportunity to also add "game" as a valid media filter in this method.

To retain functionality and avoid causing regressions,  `CPlayerCoreFactory::GetPlayers()` now supports filters. Calls that where there before (except the one in json Player.GetPlayers that I'm changing) calls it with `PLAYS_AUDIO_OR_VIDEO`. The reason for `PLAYS_AUDIO_ONLY` filter is because this method before only returned PAPlayer, despite VideoPlayer also supporting audio.

## Motivation and Context
This is part of a bigger effort to unmap player ids in the json interface from the respective playlist types. Ideally, player ids should be the unique ids stored internally in kodi by CPlayerCoreFactory. Hence, we need to have a way to get them all as a first step.

## How Has This Been Tested?
Manually, executing `jsonrpc?request={"jsonrpc":"2.0","method":"Player.GetPlayers","params":{"media":"all"},"id":2}` and the equivalent for all other media types.

**Master**
```
[
    {
        "name": "VideoPlayer",
        "playsaudio": true,
        "playsvideo": true,
        "type": "video"
    },
    {
        "name": "PAPlayer",
        "playsaudio": true,
        "playsvideo": false,
        "type": "music"
    }
]
```
**PR**
```
[
    {
        "name": "VideoPlayer",
        "playsaudio": true,
        "playsgame": false,
        "playsvideo": true,
        "type": "video"
    },
    {
        "name": "PAPlayer",
        "playsaudio": true,
        "playsgame": false,
        "playsvideo": false,
        "type": "music"
    },
    {
        "name": "RetroPlayer",
        "playsaudio": false,
        "playsgame": true,
        "playsvideo": false,
        "type": "game"
    }
]
```

@garbear and @DaveTBlake it'd be nice to have your review if possible. @Tolriq any comments from you are also appreciated.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by Tolriq** [comment]
I must admit that the motivation part frighten me a lot :)
This is a very huge part with very huge impacts and should have a proper design and target before starting to make moves.

With that said the purpose of this function is to return all compatible players to then be used later for Player.Open with playername option (https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/methods.json#L160)

This returns also UPnP players for example (And currently they can't be controlled later so that's an issue to be solved one day too :p).

So if you expose game compatible players, you need to ensure that the player.open does support games and will start them on the proper player with the proper arguments. Else this will be quite confusing to have something exposed that can't be used at all and reason of the filter.
Problem being that games currently are not exposed to JSON, so no one can list them and no one can start them.
(For the record last time an user asked me to come to help for Retroplayer I get banned for no reason, so touchy thing I suppose :))
But having game support exposed to JSON including proper controls would be nice, but should probably be done as a whole things.

TL;DR: 
1) Rewrite of players and playlist would be nice but requires proper global vision of the needs and target.
2) Exposing game players to JSON in current state of games support makes no real sense as will just confuse JSON consumers.

**Comment by enen92** [comment]
Thanks a lot for the feedback and notes on player.open (to which I was not too familiar). I agree. 
Unfortunately, after a second attempt, I'll have to give up on this. I am not seeing a bright future for Kodi.

**Comment by Tolriq** [comment]
You make me feel bad :(

Kodi just need vision and a target, the path is complex but if the target is known then building the steps is easier.

Way too many things are mixed and each things have different ways to be done depending on the entry point making every change on them risky.

Building a new central base with all things pointing at it may allow a simpler migration, but require thinking and agreement on that central thing.

**Comment by garbear** [comment]
While it may not satisfy a 100-year vision, this PR has merit in that the player core API we've currently got is incompletely exposed to JSON-RPC. I would argue for this PR by merit of completeness.

**Comment by Tolriq** [comment]
The internals to prepare are fine, if player.open does not support games, then you expose something to a consumer that can't do anything with it. If python can start games then expose to python.

Even if games are not exposed, and people would need to find the value for player.open, player.open should support the reported players by that function as both are related and dependent.

Remember that's it's a public API exposing internals that are not yet (and can't be) used may and will probably generate breaking change when it's actually needed for the other parts.

**Comment by enen92** [comment]
The internal `GetPlayers` method (not talking about the jsonrpc here) is in fact flawed since it assumes internally that only audio or music players are valid, while the callee doesn't specify any filtering. But without any callee actually needing the whole list, the change is pointless at this time (the rule has always been not to touch or change anything unless needed).

If my memory serves me well, part of Fernet's vision (kodi definitely needs more Fernets around) was to remove all the player id usage from external APIs. In fact, Python's `xbmc.Player().play()` used to support specifying a player id and that was removed and not missed at all. ApplicationPlayer follows (to the best of my knowledge) a proxy design pattern. Consumers ask for playback and kodi selects the player that best match the passed item. Accessing player properties, are also proxied to the player that is active.

The problem is that in JSONRPC player ids are mapped to content (id 0 = audio, id 1= video, id 3= Picture). This is the source of a lot of problems, not only because internally (`CPlayerCoreFactory`) the players follow a different order (0=videoplayer, 1=paplayer, 3=retroplayer, 4,... other such as external players) but also because content can change at any time. Also, video players can also play audio. So you might end up in situation while a video is playing in videoplayer, jsonrpc reports 0 as the active player, getproperties of player with id 0 assumes you're playing audio and 60% of the player properties will be empty in the response. This doesn't make any sense.
It also exposes Slideshow/picture to the same "player namespace" and that isn't even a player.

My ideia here was to start by exposing all players in `GetPlayers`, then extend it to also return the id kodi uses internally for each player, try to rework `GetActivePlayers` to always return the ID of the player that is in fact active (and not just an id based on a content that might not even be in play) and then move to Player.GetProperties and Player.GetItem. It's a fair bit of work but it's doable and matches the original vision.

Kodi starts to look a dead end and you just end up losing time and going in circles (because you'll always find something fundamentally wrong that needs to be fixed before you work on whatever you want to work with, and then you find another thing fundamentally wrong when working on your pre-condition, and so on).

As an example, this started as an attempt to solve the concurrent busydialog issue (you probably don't use addons but kodi crashes far too frequently). I am not really interested in the json-rpc. For that you need to remove GUI from VFS (plugindirectory) as a first step before even looking at issue  (I've one partial implementation in https://github.com/enen92/xbmc/commit/7cbfb0ee42903b553d14388cd2d1a5d17378a32c). In my implementation at least the call in PlaylistPlayer might be executed from the main thread and might lock the GUI. I was the one introducing that call in the first place to solve a JSON-RPC issue, so I'm looking now at alternative ways to solve the original issue without that call. Updating the playlist type after plugin resolution might break mixed playlists (whatever this is), trying to rework player id mappings to content in JSON-RPC changes the status quo of json consumers. I pretty much bet that even in the hypothetical situation that this re-work would go forward, the next blockage would be the fact that `ApplicationPlayer` is a member of Capplication and json-rpc needs to access it (atm the only way is via the `g_application` global)..

I think that at this point, the only viable "fix" for me is to revert my addition to PlaylistPlayer and reopen the original issue if I want to continue working on the plugin stuff. 
I am a reasonable guy and I'm not criticising. I am also eager to learn and to do things property instead of relying on quick hacks (even if that multiplicates the effort by several orders of magnitude). 
But unfortunately, I don't see kodi going forward in the foreseeable future without a vision nor developers to work on the different subsystems.

**Comment by Tolriq** [comment]
I do use addons and provide one that could not be in Kodi repo and was made to workaround Kodi JSON limitations as I could not made PR, when it could be used by any remote. (And faced the "nice" busydialog issue forcing me to make quite a few changes and I hate Python :p)

There's one more issue with playlist is that the slideshow part can include videos too, so it's actually still a playlist. I don't even know how this goes as in theory you can have the audio player running behind the slideshow, what happens when a video start in the slideshow when audio runs in background no idea at all :) (And this does not even take in account external players like Media Player Classic on windows or remote UPnP renderers, is remote upnp rendered playing audio while a slideshow run allowed / not allowed I think no one knows).
About mixed playlist it's about audio and video mixed in a the same playlist that can be audio or video playlist and player switching but still attached to the same playlist. I'm sure many JSON consumers does not handle this properly but due to that fact in theory consumers should ask for the playlist attached to the player (https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/types.json#L280) and not assume that if active player is audio then active playlist is audio. So in theory that hardcoding is already handled you ask for players get an id then ask the playlist, get the playlistid.
But pretty sure most do use hardcoded values. (I do for example to queue addons in the proper playlist and start proper player depending on music or video as there's no other ways)

I had and can deal with a large part of JSON changes as some of the haters seems to be gone, but those issues are at core.
First step would maybe be to decide what is allowed and what not? If GUI and Python does not allow mixed playlist and it was only possible via JSON maybe this feature should be killed to simplify near future and your short term need? (Don't know how such kind of feature drop decision are validated by the Team)
(IMO the proper future would be to have only one playlist, but this leaves the open question of slideshows with background music)

And yes the broker concept is what I talked earlier a single entry point that GUI/Python/JSON/Binary addons calls that always do the same thing in the same way the caller job being just to transform the data from a specific input mode to a broker compatible one and vice versa.

**Comment by enen92** [comment]
Having a unified interface between json, python, binary would be awesome but it is utopic at this point as all interfaces have diverged and serve different purposes. Not to even mention that different APIs have different requirements (e.g. json is supposed to be headless and not depend on GUI although it currently does, while python for example can create GUI elements). Also the unified interface as the ultimate goal won't solve fundamental issues that need to be solved first - such as the way playlists, slideshows, etc are handled in kodi that violate the single responsibility principle - so it is hardly the starting point.

**Comment by Tolriq** [comment]
You can't change fundamentals without touching all those who consume them at the same time, you can start a "broker" for each single responsibility and migrate things to it, when the public interface is common you can change implementation with a lot more ease.
Else there's never a starting point as no one have the necessary time and knowledge to do all at once.

Each Internal/Json/Python/GUI/Binary serves different purpose and clients but none should access directly the players, none should access directly the database and so on.
Nothing prevents the definition of a proper interface for the player that all should use to query status and play stuff. Migrate those to that interface then change the internals much more easily.

As all is directly accessing all there's not a lot of ways out other than adding and migrating to new abstraction layers that will after allows implementation rewrites.

Still a complex and hard path, but possible, if you effectively want to start by migrating the core without adding the abstraction layers then effectively there's nearly no chance it happens.
